### PR TITLE
Add weakMapMemoize utility.

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

Introduces a memoize function thats more powerful than what lodash offers. Lodash memoize only works with string keys, and so it requires you to take your arguments and somehow serialize it to a string. Depending on the object this can be very slow if the object is big. To avoid the headache of serialization I'm adding a weakMapMemoize which is much more flexible and can cache any function mixing both primitives and objects. 

Going to use this to memoize the `filterByQuery` helpers for the GanttChart since it looks like we call the utility in a lot of different places separately and having them coordinate so that there's only 1 call is a bit tricky.

## How I Tested These Changes

jest tests

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/util/weakMapMemoize.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    // Compute the result
+    const result = fn.apply(this, args);
+
+    // Cache the result
+    currentCache.result = result;
+
+    // If LRU cache is enabled, manage the cache entries
+    if (lruCache) {
+      const cacheEntryKey: any[] = path.slice(); // Clone the path to ensure uniqueness
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/util/weakMapMemoize.ts:125]: oops, gonna just generate a unique ID here

Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/util/weakMapMemoize.ts:125]: nvm, this is fine, i remember now, this is already unique

Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/util/weakMapMemoize.ts:125]: Though maybe I can just use `Symbol()` to avoid copying an array...

**Review by hellendag — APPROVED** [review]
Cool, seems reasonable to me.

**Pull request merged**